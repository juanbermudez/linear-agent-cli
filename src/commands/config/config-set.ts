import { Command } from "@cliffy/command"
import { getConfigManager } from "../../utils/config-manager.ts"
import { error as errorColor, success as successColor } from "../../utils/styling.ts"

interface SetOptions {
  global?: boolean
  json?: boolean
  format?: string
}

export const setCommand = new Command()
  .name("set")
  .description("Set a configuration value")
  .arguments("<key:string> <value:string>")
  .option("-g, --global", "Set in global config (not yet implemented)")
  .option("-j, --json", "Output result as JSON")
  .option("--format <format:string>", "Output format: text|json")
  .action(async (options: SetOptions, key: string, value: string) => {
    const useJson = options.json || options.format === "json"

    try {
      const configManager = await getConfigManager()

      // Parse value - handle booleans and numbers
      let parsedValue: unknown = value
      if (value === "true") {
        parsedValue = true
      } else if (value === "false") {
        parsedValue = false
      } else if (/^\d+$/.test(value)) {
        parsedValue = parseInt(value, 10)
      } else if (/^\d+\.\d+$/.test(value)) {
        parsedValue = parseFloat(value)
      }

      configManager.set(key, parsedValue)
      await configManager.save()

      if (useJson) {
        console.log(
          JSON.stringify({
            success: true,
            operation: "set",
            key,
            value: parsedValue,
          }, null, 2),
        )
      } else {
        console.log(successColor(`âœ“ Set ${key} = ${value}`))
      }
    } catch (err) {
      const error = err as Error
      if (useJson) {
        console.error(
          JSON.stringify({
            success: false,
            error: {
              code: "API_ERROR",
              message: error.message,
            },
          }, null, 2),
        )
      } else {
        console.error(errorColor(`Error: ${error.message}`))
      }
      Deno.exit(1)
    }
  })
