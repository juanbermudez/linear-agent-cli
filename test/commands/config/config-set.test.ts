import { snapshotTest } from "@cliffy/testing"
import { setCommand } from "../../../src/commands/config/config-set.ts"
import { commonDenoArgs } from "../../utils/test-helpers.ts"
import { join } from "@std/path"

// Helper to create a temporary directory with config file
async function setupTempConfig(): Promise<{ tempDir: string; cleanup: () => Promise<void> }> {
  const tempDir = await Deno.makeTempDir()
  const configPath = join(tempDir, "linear.toml")

  // Create initial config
  await Deno.writeTextFile(configPath, `workspace = "test-workspace"\nteam_id = "ENG"\n`)

  // Change to temp directory so config is found
  const originalDir = Deno.cwd()
  Deno.chdir(tempDir)

  return {
    tempDir,
    cleanup: async () => {
      Deno.chdir(originalDir)
      await Deno.remove(tempDir, { recursive: true })
    },
  }
}

// Test help output
await snapshotTest({
  name: "Config Set Command - Help Text",
  meta: import.meta,
  colors: false,
  args: ["--help"],
  denoArgs: commonDenoArgs,
  async fn() {
    await setCommand.parse()
  },
})

// Test setting a simple value
await snapshotTest({
  name: "Config Set Command - Simple Value",
  meta: import.meta,
  colors: false,
  args: ["workspace", "new-workspace", "--json"],
  denoArgs: commonDenoArgs,
  async fn() {
    const { cleanup } = await setupTempConfig()

    try {
      await setCommand.parse()
    } finally {
      await cleanup()
    }
  },
})

// Test setting a nested value with dot notation
await snapshotTest({
  name: "Config Set Command - Nested Value (Dot Notation)",
  meta: import.meta,
  colors: false,
  args: ["defaults.project.status", "In Progress", "--json"],
  denoArgs: commonDenoArgs,
  async fn() {
    const { cleanup } = await setupTempConfig()

    try {
      await setCommand.parse()
    } finally {
      await cleanup()
    }
  },
})

// Test setting a boolean value
await snapshotTest({
  name: "Config Set Command - Boolean Value",
  meta: import.meta,
  colors: false,
  args: ["interactive.enabled", "true", "--json"],
  denoArgs: commonDenoArgs,
  async fn() {
    const { cleanup } = await setupTempConfig()

    try {
      await setCommand.parse()
    } finally {
      await cleanup()
    }
  },
})

// Test setting a number value
await snapshotTest({
  name: "Config Set Command - Number Value",
  meta: import.meta,
  colors: false,
  args: ["defaults.milestone.targetDateOffset", "30", "--json"],
  denoArgs: commonDenoArgs,
  async fn() {
    const { cleanup } = await setupTempConfig()

    try {
      await setCommand.parse()
    } finally {
      await cleanup()
    }
  },
})

// Test setting a deeply nested value
await snapshotTest({
  name: "Config Set Command - Deeply Nested Value",
  meta: import.meta,
  colors: false,
  args: ["level1.level2.level3.value", "deep", "--json"],
  denoArgs: commonDenoArgs,
  async fn() {
    const { cleanup } = await setupTempConfig()

    try {
      await setCommand.parse()
    } finally {
      await cleanup()
    }
  },
})

// Test non-JSON output format
await snapshotTest({
  name: "Config Set Command - Text Output",
  meta: import.meta,
  colors: false,
  args: ["team_id", "PLATFORM"],
  denoArgs: commonDenoArgs,
  async fn() {
    const { cleanup } = await setupTempConfig()

    try {
      await setCommand.parse()
    } finally {
      await cleanup()
    }
  },
})

// Test setting with string that looks like a number but has leading zero
await snapshotTest({
  name: "Config Set Command - String With Leading Zero",
  meta: import.meta,
  colors: false,
  args: ["code", "0123", "--json"],
  denoArgs: commonDenoArgs,
  async fn() {
    const { cleanup } = await setupTempConfig()

    try {
      await setCommand.parse()
    } finally {
      await cleanup()
    }
  },
})

// Test setting false boolean value
await snapshotTest({
  name: "Config Set Command - False Boolean",
  meta: import.meta,
  colors: false,
  args: ["output.color", "false", "--json"],
  denoArgs: commonDenoArgs,
  async fn() {
    const { cleanup } = await setupTempConfig()

    try {
      await setCommand.parse()
    } finally {
      await cleanup()
    }
  },
})

// Test setting a float value
await snapshotTest({
  name: "Config Set Command - Float Value",
  meta: import.meta,
  colors: false,
  args: ["version", "1.5", "--json"],
  denoArgs: commonDenoArgs,
  async fn() {
    const { cleanup } = await setupTempConfig()

    try {
      await setCommand.parse()
    } finally {
      await cleanup()
    }
  },
})
