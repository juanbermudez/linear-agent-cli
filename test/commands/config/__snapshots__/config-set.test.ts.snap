export const snapshot = {};

snapshot[`Config Set Command - Help Text 1`] = `
stdout:
"
Usage: set <key> <value>

Description:

  Set a configuration value

Options:

  -h, --help              - Show this help.                             
  -g, --global            - Set in global config (not yet implemented)  
  -j, --json              - Output result as JSON                       
  --format      <format>  - Output format: text|json                    

"
stderr:
""
`;

snapshot[`Config Set Command - Simple Value 1`] = `
stdout:
'{
  "success": true,
  "operation": "set",
  "key": "workspace",
  "value": "new-workspace"
}
'
stderr:
""
`;

snapshot[`Config Set Command - Nested Value (Dot Notation) 1`] = `
stdout:
'{
  "success": true,
  "operation": "set",
  "key": "defaults.project.status",
  "value": "In Progress"
}
'
stderr:
""
`;

snapshot[`Config Set Command - Boolean Value 1`] = `
stdout:
'{
  "success": true,
  "operation": "set",
  "key": "interactive.enabled",
  "value": true
}
'
stderr:
""
`;

snapshot[`Config Set Command - Number Value 1`] = `
stdout:
'{
  "success": true,
  "operation": "set",
  "key": "defaults.milestone.targetDateOffset",
  "value": 30
}
'
stderr:
""
`;

snapshot[`Config Set Command - Deeply Nested Value 1`] = `
stdout:
'{
  "success": true,
  "operation": "set",
  "key": "level1.level2.level3.value",
  "value": "deep"
}
'
stderr:
""
`;

snapshot[`Config Set Command - Text Output 1`] = `
stdout:
"âœ“ Set team_id = PLATFORM
"
stderr:
""
`;

snapshot[`Config Set Command - String With Leading Zero 1`] = `
stdout:
'{
  "success": true,
  "operation": "set",
  "key": "code",
  "value": 123
}
'
stderr:
""
`;

snapshot[`Config Set Command - False Boolean 1`] = `
stdout:
'{
  "success": true,
  "operation": "set",
  "key": "output.color",
  "value": false
}
'
stderr:
""
`;

snapshot[`Config Set Command - Float Value 1`] = `
stdout:
'{
  "success": true,
  "operation": "set",
  "key": "version",
  "value": 1.5
}
'
stderr:
""
`;
